scammer-label-propagation/src/agent.py

```diff
@@ -0,0 +1,149 @@
+
+def provide_handle_alert(w3):
+    logger.debug("provide_handle_alert called")
+    
+
+    def handle_alert(alert_event: forta_agent.alert_event.AlertEvent) -> list:
+        logger.debug("handle_alert inner called")
+        logger.debug(f"AlertId:{alert_event.alert_id};\tName:{alert_event.name};\tAlertHash:{alert_event.hash}")
+        t = time.time()
+        global executor
+        global addresses_analyzed
+        global global_futures
+
+        list_of_addresses = []
+        if alert_event.alert.alert_id == 'SCAM-DETECTOR-ADDRESS-POISONING':
+            logger.info(f"Address poisoning. Addresses: {';'.join([label.entity for label in alert_event.alert.labels])}")
+            return []
+        for label in alert_event.alert.labels:
+            if label.confidence >= ATTACKER_CONFIDENCE and label.entity_type == forta_agent.EntityType.Address:
+                list_of_addresses.append(label.entity)
+        list_of_addresses = list(set(list_of_addresses))
+        for address in list_of_addresses:
+            if address not in addresses_analyzed:
+                logger.debug(f"Adding address {address} to the pool")
+                global_futures[address] = executor.submit(run_all_extended, address, alert_event.alert.alert_id)
+                addresses_analyzed.append(address)
+        logger.info(f"Alert {alert_event.alert.alert_id}:\t{time.time() - t:.10f} s. {len(list_of_addresses)} addresses: {';'.join(list_of_addresses)}")
+        return []
+
+    return handle_alert
+
+web3 = Web3(Web3.HTTPProvider(get_json_rpc_url()))
+
+real_handle_alert = provide_handle_alert(web3)
+
+
+def handle_alert(alert_event: forta_agent.alert_event.AlertEvent) -> list:
+    logger.debug("handle_alert called")
+    return real_handle_alert(alert_event)
+
+
+def provide_handle_block(w3):
+    logger.debug("provide_handle_block called")
+
+    def handle_block(block_event) -> list:
+        logger.debug("handle_block inner called")
+        t = time.time()
+        global global_futures
+        global global_alerts
+
+        completed_futures = []
+        running_futures = 0
+        pending_futures = 0
+        for address, future in global_futures.items():
+            if future.running():
+                running_futures += 1
+            elif future.done():
+                try:
+                    global_alerts += future.result()
+                    completed_futures.append(address)
+                except Exception as e:
+                    logger.error(f"Exception {e} occurred while collecting results from address {address}")
+            else:
+                pending_futures += 1
+        for address in completed_futures:
+            global_futures.pop(address)
+        # We return the first MAX_FINDINGS findings, and remove them from the list. Otherwise
+        # we cache them in global alerts and will return them in the next block
+        alerts = global_alerts[:MAX_FINDINGS]
+        global_alerts = global_alerts[MAX_FINDINGS:]
```

Comment:
you probably want to add some local persistence; what if your bot restarts or crashes? Any alerts would be lost.

Reply:
I want to implement global persistence instead of local, then it will also be compatible with sharding